File: 2019/Day 06/sym.rb
$sys = Hash.new; require 'set'
File.read("input.txt").split.each do
  |ln| pr, ch = ln.split(')')
$sys[ch] = pr end
# Iterate Backwards
def flatten(x)
  Set.new(Enumerator.new do
    |en| while $sys.key? x
      en.yield (x = $sys[x])
  end end)
end
# Such a nice day
puts "Silver: #{$sys.keys.map{|x| flatten(x).size}.sum}",
  "Gold: #{(flatten('SAN') ^ flatten('YOU')).size}"

File: 2019/Day 06/sym.py
Graph = {Ch: Pr for Pr, Ch in
    (ln.strip().split(')') for ln
in open("2019/Day 06/input.txt"))}
# Iterate Backwards
def Flatten(x):
    while x in Graph:
        x = Graph[x]
        yield x
# Part 1
print("Silver:",
    sum(len((*Flatten(x),))
for x in Graph))
# Part 2
print("Gold:",
    len({*Flatten('SAN')}
^ {*Flatten('YOU')}))
